Update draft-ietf-cats-usecases-requirements.md (#2)
cats-wg/draft-ietf-cats-usecases-requirements: draft-ietf-cats-usecases-requirements.md

   This document makes use of the following terms:

>   Client:

Reviewer: Add a mention that the terminology echoes what is draft-ldbc-cats-framework.

Reviewer 2: I wonder if we need to define the teminology in this document and change the framework draft to reference this one?

Reviewer: I prefer to do it in the other way around (that is, leave the terms in the framework). We need to keep in mind this part from the charter:

> o Groundwork may be documented via a set of informational Internet-
> Drafts, **not necessarily for publication as RFCs**:

Reviewer 2: You are right.

     An endpoint that is connected to a service provider network.

   Computing-Aware Traffic Steering (CATS):  A traffic engineering

>      approach [I-D.ietf-teas-rfc3272bis] that takes into account the

Reviewer: ```suggestion
     approach {{?I-D.ietf-teas-rfc3272bis}} that takes into account the
```

     dynamic nature of computing resources and network state to optimize
     service-specific traffic forwarding towards a given service contact
     instance.  Various relevant metrics may be used to enforce such
     computing-aware traffic steering policies.

   Service:  An offering that is made available by a provider by
     orchestrating a set of resources (networking, compute, storage,
     etc.).  Which and how these resources are solicited is part of the
     service logic which is internal to the provider.  For example,
     these resources may be:
 
        * Exposed by one or multiple processes (a.k.a.  Service

>         Functions (SFs) ).  [RFC7665]

Reviewer: ```suggestion
        Functions (SFs)) {{?RFC7665}}.
```

 
        * Provided by virtual instances, physical, or a combination
        thereof.

        * Hosted within the same or distinct nodes.

        * Hosted within the same or multiple service sites.

        * Chained to provide a service using a variety of means.

        How a service is structured is out of the scope of CATS.

        The same service can be provided in many locations; each of them

>         constitutes a service instance.

Reviewer: ```suggestion
        constitutes a service instance.
        
Service contact instance:
:  A client-facing service function instance
      that is responsible for receiving requests in the context of a
      given service.  A service request is processed according to the
      service logic (e.g., handle locally or solicit backend resources).
      Steering beyond the service contact instance is hidden to both
      clients and CATS components.

 : A service contact instance is reachable via at least one Egress
      CATS Forwarder.

: A service can be accessed via multiple service contact instances
      running at the same or different locations (service sites).

: The same service contact instance may dispatch service requests to
      one or more service instances (e.g., an instance that behaves as a
      service load-balancer).
```

Reviewer 3: If both service instance and service contact instance exist, from my understanding, service instance is not a very useful concept. In the document, there are a number of occurences of "selection of service or service selection" here and there. After introducing service contact instance, which is selected, service instance or service contact instance? More like service contact instance to me.

Reviewer 2: s/service instance/service contact instance/g

Reviewer: > If both service instance and service contact instance exist, from my understanding, service instance is not a very useful concept. In the document, there are a number of occurences of "selection of service or service selection" here and there. After introducing service contact instance, which is selected, service instance or service contact instance? More like service contact instance to me.

Agree. service instance should be used in very few occurrences; what matters is the service contact instance.

Reviewer: > If both service instance and service contact instance exist, from my understanding, service instance is not a very useful concept. In the document, there are a number of occurences of "selection of service or service selection" here and there. After introducing service contact instance, which is selected, service instance or service contact instance? More like service contact instance to me.

Agree. service instance should be used in very few occurrences; what matters is the service contact instance.

Reviewer 4: Hi Med, my understanding is that these terminologies should echo what is in draft-ldbc-cats-framework, but the problem is that not all terms are needed in this use case & requirement draft. The definition of "service contact instance" will introduce other terms like Egress, CATS forward, etc, which are not referenced in this document. That's why I didn't refer these terms from draft-ldbc-cats-framework. And BTW, currently, draft-ldbc-cats-framework is not adopted by the WG, and terminologies within the draft may be updated. What do you think?

   towards compute capabilities and resources, and their corelation with
   services.  Therefore, a desirable system:

>    R1: MUST provide a discovery and resolving methodology for the

Reviewer: I would maintain the old formatting.

```suggestion
   R1:
   : MUST provide a discovery and resolving methodology for the
```

   others {{TR22.874}}.  Of course, specific metrics might not be utilized
   at all in other scenarios.  Hence:

   R12: In addition to common metrics that are agreed by all CATS
   components like processing delay, there SHOULD be some ways for

>    metrics definition in different dimensions, which is used for the

Reviewer: not sure what is meant by "dimensions" here.

Reviewer 3: ```suggestion
   allowing the customized metrics definition, which is used for the
```

Reviewer 5: what are 'customized definitions'? Adding one's own definition outside the common ones? If so, how is a decision to be made then for instance (contact point) selection since the selection point would need to know and understand that 'customized' definition? Are we opening a can of worms here?

Reviewer 3: Vendor specific instead of customized? From my understanding, the original text tried to explain there were different ways to express metrics. Some can be standardized without much controversies (called common metrics in the given text), and some may not.

Reviewer 5: I understand that ;-) But how do we want to design an interoperable system (since this is an SDO work) where the metrics are vendor-specific, which means that the decision algorithms are likely, too. What is left is the encapsulation format and that's it? Sorry for sounding cynical but opening the door to vendor-specific metrics renders the idea of thinking about metrics rather futile but it postulates the idea of vendor lock-in (for the metrics and the decision algorithms) for the deployed solution to work.

Reviewer 5: BTW, an approach where 'vendor-specific- would work is to define an interface of conveying metrics to the decision algorithms that does not, well, require ANY metric knowledge at all. My suggestion to use mere cardinals at the interface allows for interpreting those cardinals in any way you want at the application and vendor level. You may now have energy-aware decision (e.g., the cardinal being a Joule expression of your instance), compute-aware decision (e.g., the cardinal reflecting a deployed compute unit or even an instantaneous compute load), or a storage-aware decision (e.g., the cardinal expressing the occupancy of the associated hard disk). It can be whatever the application and vendor wants while the interface, conveying just a cardinal, remains the same. But having an interface where we allow for agreed/common metrics and proprietary ones at the same time creeps the vendor lock into the backdoor

Reviewer 2: > BTW, an approach where 'vendor-specific- would work is to define an interface of conveying metrics to the decision algorithms that does not, well, require ANY metric knowledge at all. My suggestion to use mere cardinals at the interface allows for interpreting those cardinals in any way you want at the application and vendor level. You may now have energy-aware decision (e.g., the cardinal being a Joule expression of your instance), compute-aware decision (e.g., the cardinal reflecting a deployed compute unit or even an instantaneous compute load), or a storage-aware decision (e.g., the cardinal expressing the occupancy of the associated hard disk). It can be whatever the application and vendor wants while the interface, conveying just a cardinal, remains the same. But having an interface where we allow for agreed/common metrics and proprietary ones at the same time creeps the vendor lock into the backdoor

What is the difference? You still get proprietary interpretation of the same number which make it hard to interop, right?

Reviewer 5: Not really. If the cardinal is input to a simple utility function, the realization of it in the intermediary elements as well as the signaling of values can all be standard-based, which is what you want. What remains proprietary, as you point out, is the 'interpretation' of the cardinal and its optimization (through the utility function) - but that is OK since it's entirely left to the application to interpret it anyway, isn't it?

Reviewer 5: ...I mean why would/should the overlay provider know anyway on which basis the application would want to optimize its reception of traffic at different egress points? Hence, the more semantics are exposed across the business interface (which exists between app and CATS provider), the more issues you will have in terms of acceptance.

Reviewer 4: Hi, all, due to time limitation for uploading the draft, I just rewording this sentence to be "...there SHOULD be some other ways for metrics definition, which is used for the selection of specific service instance." I know this is not the perfect requirement, and it needs further discussion.

   transfer (e.g., when metrics change significantly).  So in those
   situations:

>    R15: session as well as service continuity MUST be maintained when

Reviewer: I don't parse this requirement

Reviewer 3: same question here. Shall we just say service continuity?

Reviewer 5: Isn't session continuity that is important, i.e., the instance affinity?

Reviewer 3: It looks like session affitinity/persistence, instance affinity, and service continuity(?) were used quite interchangeably in the document. like: 
"Such required transfer of state/context makes it desirable to have session persistence (or instance affinity) as the default, ..."  
Using a single terminology?

Reviewer 5: Agree on that! Given that CATS is about selecting the right instance contact point, I think we should converge to use 'instance affinity', i.e., the affinity of one needed decision at time t to a previous decision at time t-eps.

Reviewer 2: Agree. This should be aligned with section 4.5 of [CATS framework draft](https://datatracker.ietf.org/doc/draft-ldbc-cats-framework/) as well which uses service contact instance affinity... But it seems like the service contact instance is defined in the framework draft instead of this one.
